File: ProjectCS/GroupTable/TXTReportExporter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace SessionResultsCS
{
    public class TXTReportExporter : FileReportExporter
    {
        public ITextTableRenderer Renderer { get; set; }


        public TXTReportExporter(string filePath, ITextTableRenderer renderer, string title = "", string body = "", bool includeDate = false)
            : base(filePath, title, body, includeDate)
        {
            Renderer = renderer;
        }


        public override void Export(GroupTable groupTable)
        {
            GroupTableData data = groupTable.GetTableData();
            
            using (StreamWriter writer = new StreamWriter(FilePath))
            {
                if (Title != "")
                {
                    writer.WriteLine(Title);
                    writer.WriteLine();
                }
                if (Body != "")
                {
                    writer.WriteLine(Body);
                    writer.WriteLine();
                }
                if (IncludeDate)
                {
                    writer.WriteLine($"Дата генерации: {DateTime.Today.ToShortDateString()}");
                    writer.WriteLine();
                }

                writer.WriteLine(Renderer.RenderTable(data));
                writer.WriteLine();

                for (int i = 0; i < data.Disciplines.Length; i++)
                {
                    writer.WriteLine($"{i + 1} - {data.Disciplines[i].Name}");
                }
            }
        }
    }
}

File: Project/GroupTable/TXTReportExporter.cpp
#include "GroupTable/TXTReportExporter.hpp"

#include <fstream>
#include <chrono>
#include <iomanip>
#include <sstream>


TXTReportExporter::TXTReportExporter(const std::string& filePath, std::unique_ptr<ITextTableRenderer> renderer, const std::string& title, const std::string& body, bool includeDate) :
	FileReportExporter(filePath, title, body, includeDate), m_renderer(std::move(renderer)) {}

TXTReportExporter::TXTReportExporter(const TXTReportExporter& other) :
	FileReportExporter(other) {}

void TXTReportExporter::Export(const GroupTable& groupTable) {
	GroupTableData data = groupTable.GetTableData();
	

	std::ofstream fout(m_filePath);

	if (!m_title.empty()) {
		fout << m_title << std::endl << std::endl;
	}
	if (!m_body.empty()) {
		fout << m_body << std::endl << std::endl;
	}
	if (m_includeDate) {
		time_t now = time(0);
		std::tm tmNow;
		localtime_s(&tmNow, &now);
		fout << "Дата генерации: " << std::put_time(&tmNow, "%d.%m.%Y") << std::endl << std::endl;
	}

	fout << m_renderer->RenderTable(data) << std::endl;
	fout << std::endl;

	for (int i = 0; i < data.GetDisciplines().size(); i++) {
		fout << (i+1) << " - " << data.GetDisciplines()[i]->GetName() << std::endl;
	}
}
